File: PROCESS SCHEDULING/CPP/FCFS.cpp
#include <stdio.h>
#include <stdlib.h>
struct process {
    int pid, at, bt, ct, st, rt;
    float tat, wt;
};
typedef struct process process;
int compareAT(const void *p1, const void *p2) {
    int a = ((process *)p1)->at;
    int b = ((process *)p2)->at;
    return (a == b) ? 0 : (a < b) ? -1 : 1;
}
int compareST(const void *p1, const void *p2) {
    int st1 = ((process *)p1)->st;
    int st2 = ((process *)p2)->st;
    return (st1 == st2) ? 0 : (st1 < st2) ? -1 : 1;
}
int compareId(const void *p1,const void *p2){
	int a=((process *)p1)->pid;
	int b=((process *)p2)->pid;
	return a-b;
}
int main() {
    int n;
    printf("Enter number of processes: ");
    scanf("%d", &n);
    process p[n];
    // Input arrival and burst times for each process
    for (int i = 0; i < n; i++) {
        printf("\nEnter arrival time for process P%d: ", i + 1);
        scanf("%d", &p[i].at);
        printf("Enter burst time for process P%d: ", i + 1);
        scanf("%d", &p[i].bt);
        p[i].pid = i + 1;
    }
    // Sort processes by arrival time
    qsort((void *)p, n, sizeof(process), compareAT);
    float totalBt = 0.0, totalWt = 0.0, totalTat = 0.0;
    // Calculate completion time, turnaround time, waiting time, etc.
    for (int i = 0; i < n; i++) {
        if (i == 0) {
            p[i].st = p[i].at;
        } else {
            p[i].st = (p[i].at > p[i - 1].ct) ? p[i].at : p[i - 1].ct;
        }
        p[i].ct = p[i].st + p[i].bt;
        p[i].tat = p[i].ct - p[i].at;
        p[i].wt = p[i].tat - p[i].bt;
        p[i].rt = p[i].wt;

        totalBt += p[i].bt;
        totalWt += p[i].wt;
        totalTat += p[i].tat;
    }
    // Calculate averages and CPU utilization
    float avgWt = totalWt / n;
    float avgTat = totalTat / n;
    float endTime = 0;
    for (int i = 0; i < n; i++) {
        if (p[i].ct > endTime) {
            endTime = p[i].ct;
        }
    }
    float cpu_utilization = (totalBt / endTime) * 100;
    float throughput = n / endTime;
    // Print Gantt chart sequence (only process IDs in execution order)
    qsort((void *)p, n, sizeof(process), compareST);
    printf("\nGantt Chart Process Sequence:\n|");
    for (int i = 0; i < n; i++) {
        printf("  P%d  |", p[i].pid);
    }
    printf("\n");
    // Print process details sorted by PID
    qsort((void *)p,n,sizeof(process),compareId);
    printf("\nPID\tAT\tBT\tST\tCT\tTAT\t\tWT\t\tRT\n");
    for (int i = 0; i < n; i++) {
        printf("P%d\t%d\t%d\t%d\t%d\t%.2f\t\t%.2f\t\t%d\n", 
            p[i].pid, p[i].at, p[i].bt, p[i].st, p[i].ct, p[i].tat, p[i].wt, p[i].rt);
    }
    // Print totals and averages
    printf("\nTotal Turn Around Time: %.2f\nAverage Turn Around Time: %.2f\n", totalTat, avgTat);
    printf("Total Waiting Time: %.2f\nAverage Waiting Time: %.2f\n", totalWt, avgWt);
    printf("CPU Utilization: %.2f%%\nThroughput: %.2f processes/unit time\n", cpu_utilization, throughput);

    return 0;
}

File: PROCESS SCHEDULING/JS/FCFS.js
class Process {
  constructor(pid, at, bt) {
    this.pid = pid;
    this.at = at;
    this.bt = bt;
    this.st = 0;
    this.ct = 0;
    this.tat = 0;
    this.wt = 0;
    this.rt = 0;
  }
}

// Sample Input (you can take this via prompt or form)
const n = 3;
const p = [
  new Process(1, 0, 5),
  new Process(2, 2, 3),
  new Process(3, 4, 1)
];

// Sort by arrival time
p.sort((a, b) => a.at - b.at);

let totalBt = 0, totalWt = 0, totalTat = 0;

// Calculations
for (let i = 0; i < n; i++) {
  if (i === 0) {
    p[i].st = p[i].at;
  } else {
    p[i].st = Math.max(p[i].at, p[i - 1].ct);
  }
  p[i].ct = p[i].st + p[i].bt;
  p[i].tat = p[i].ct - p[i].at;
  p[i].wt = p[i].tat - p[i].bt;
  p[i].rt = p[i].wt;

  totalBt += p[i].bt;
  totalWt += p[i].wt;
  totalTat += p[i].tat;
}

const avgWt = totalWt / n;
const avgTat = totalTat / n;
let endTime = Math.max(...p.map(proc => proc.ct));
const cpuUtilization = (totalBt / endTime) * 100;
const throughput = n / endTime;

// Gantt Chart Order
p.sort((a, b) => a.st - b.st);
console.log("\nGantt Chart:");
console.log(p.map(proc => `P${proc.pid}`).join(" -> "));

// Restore original order by PID
p.sort((a, b) => a.pid - b.pid);

// Display Table
console.log("\nPID\tAT\tBT\tST\tCT\tTAT\tWT\tRT");
p.forEach(proc => {
  console.log(`P${proc.pid}\t${proc.at}\t${proc.bt}\t${proc.st}\t${proc.ct}\t${proc.tat.toFixed(2)}\t${proc.wt.toFixed(2)}\t${proc.rt}`);
});

// Averages
console.log(`\nTotal Turn Around Time: ${totalTat.toFixed(2)}\nAverage Turn Around Time: ${avgTat.toFixed(2)}`);
console.log(`Total Waiting Time: ${totalWt.toFixed(2)}\nAverage Waiting Time: ${avgWt.toFixed(2)}`);
console.log(`CPU Utilization: ${cpuUtilization.toFixed(2)}%\nThroughput: ${throughput.toFixed(2)} processes/unit time`);
